prometheus/metrics.go

```diff
@@ -18,239 +17,188 @@ var (
+		fetchRange(api, query, bounds, avgChan)
 	}()
 
 	// Median
 	go func() {
 		query := fmt.Sprintf(
-			"histogram_quantile(0.5, sum(rate(%s_bucket%s[%s])) by (le))", metricName, labels, duration)
-		med, err := fetchScalarDouble(api, query, now)
-		if err != nil {
-			errChan <- err
-		} else {
-			medChan <- med
-		}
+			"histogram_quantile(0.5, sum(rate(%s_bucket%s[%s])) by (le))", metricName, labels, rateInterval)
+		fetchRange(api, query, bounds, medChan)
 	}()
 
 	// Quantile 95
 	go func() {
 		query := fmt.Sprintf(
-			"histogram_quantile(0.95, sum(rate(%s_bucket%s[%s])) by (le))", metricName, labels, duration)
-		p95, err := fetchScalarDouble(api, query, now)
-		if err != nil {
-			errChan <- err
-		} else {
-			p95Chan <- p95
-		}
+			"histogram_quantile(0.95, sum(rate(%s_bucket%s[%s])) by (le))", metricName, labels, rateInterval)
+		fetchRange(api, query, bounds, p95Chan)
 	}()
 
 	// Quantile 99
 	go func() {
 		query := fmt.Sprintf(
-			"histogram_quantile(0.99, sum(rate(%s_bucket%s[%s])) by (le))", metricName, labels, duration)
-		p99, err := fetchScalarDouble(api, query, now)
-		if err != nil {
-			errChan <- err
-		} else {
-			p99Chan <- p99
-		}
+			"histogram_quantile(0.99, sum(rate(%s_bucket%s[%s])) by (le))", metricName, labels, rateInterval)
+		fetchRange(api, query, bounds, p99Chan)
 	}()
 
-	var histo = MetricHistogram{found: true}
-	var metricError error
-	select {
-	case histo.Average = <-avgChan:
-	case metricError = <-errChan:
-	}
-	select {
-	case histo.Median = <-medChan:
-	case metricError = <-errChan:
-	}
-	select {
-	case histo.Percentile95 = <-p95Chan:
-	case metricError = <-errChan:
-	}
-	select {
-	case histo.Percentile99 = <-p99Chan:
-	case metricError = <-errChan:
+	ch <- Histogram{
+		Average:      <-avgChan,
+		Median:       <-medChan,
+		Percentile95: <-p95Chan,
+		Percentile99: <-p99Chan}
+}
+
+func fetchTimestamp(api v1.API, query string, t time.Time, ch chan vectorResult) {
+	result, err := api.Query(context.Background(), query, t)
+	if err != nil {
+		ch <- vectorResult{err: err}
+		return
 	}
-	// Any NaN value is considered as a missing metric
-	if math.IsNaN(histo.Average) || math.IsNaN(histo.Median) || math.IsNaN(histo.Percentile95) || math.IsNaN(histo.Percentile99) {
-		return &MetricHistogram{found: false}, nil
+	switch result.Type() {
+	case model.ValVector:
+		ch <- vectorResult{v: result.(model.Vector)}
+		return
 	}
-	return &histo, metricError
+	ch <- vectorResult{err: fmt.Errorf("Invalid query, vector expected: %s", query)}
 }
 
-func fetchScalarDouble(api v1.API, query string, now time.Time) (float64, error) {
-	result, err := api.Query(context.Background(), query, now)
+func fetchRange(api v1.API, query string, bounds v1.Range, ch chan Metric) {
+	result, err := api.QueryRange(context.Background(), query, bounds)
 	if err != nil {
-		return 0, err
+		ch <- Metric{err: err}
```

Comment:
Minor thing.
Adding the error in the Metric type is intended ?
I see we use the vectorResult one as internal one to the package and now the Metric.
Probably yes, but just to confirm.

Reply:
`vectorResult` and `Metric` are a bit similar but they don't hold the same kind of data (the first provides 1 datapoint, the other is a timeseries) so we have to create different types ... and `vectorResult` is not exported because it is a transitional data used to build `Health`.

We could decide to create `matrixResult` similar to `vectorResult` but I just took a shortcut to avoid too many types & conversions between them. Maybe it's less clean but simpler.